### ISSUE TITLE
Input-safe

### ISSUE THREAD

**Issue by Ditiae** [NightShadeI/Revolution-Rotation-Calculator-#7]
![image](https://user-images.githubusercontent.com/30428646/32671543-75e7f6da-c60d-11e7-97f7-ca22d1105b30.png)

What's the point of the "What style will you be using?" line? And should it be type safe and check for "Melee," "Ranged," "Magic," "Mage," or other inputs. At the moment I'm not sure what it does, nor do I know if it's necessary, but I don't think "Your mom" is gonna be too helpful. 😛 
That goes for the other lines as well.

**Comment by EZtouch** [comment]
"Different styles of combat tree give varying amounts of adrenaline from auto attacks".

I was thinking that the prompt itself should be removed and the combat style used should be the same as the first selected ability. If defensive abilities are added, this wouldn't work. Also, there are "hybrid" pieces of armor; there might be hybrid weapons in the game 🤔. If there are/will be, people might be using multiple attack styles, but then the game algorithm for adrenaline gain will (probably) change too.
https://www.reddit.com/r/runescape/comments/498rbc/hybrid_weapons_invention/

To add to that, some people have their action bar setup as 2-in-1, 7 melee abilities then 7 ranged abilities, simply switching the weapon would cause the other half of the bar to be used

As for me, I have always used "crush", "stab", "slash" when prompted, perhaps we should limit the user input to a set of keyword 🙂

**Comment by Ditiae** [comment]
Or give an example of what type of input you are looking for in the prompt?

**Comment by EZtouch** [comment]
The easiest fix would be changing it to "Are you using magic combat style (y/n)".
The same goes for attack speed prompt, it should have the expected input in brackets too. 🙂

**Comment by NightShadeI** [comment]
Yeah this is a good point, I will see if I can get around to changing this in a bit. Bad practice too.

**Comment by NightShadeI** [comment]
Alright, been thinking about it, and for this there are 3 fixes, let me know what you think:

(1) Simply change validation methods and wording of inputs - won't take long at all

(2) Create some kind of file (perhaps configurations.txt) or something similar - this makes it much easier for the user, as common options and abilities can be left the same, and any mistakes can easily be corrected, however more sophisticated validation methods must be used so will take a while to implement.

(3) Create a UI to collect input, this allows for self-validation and of course is much more tidy for  the user. However, python is a console based language and I am not so used to programming with UI's so either someone will have to implement the change themselves, or I will have to spend a while learning a module such as Tkinter.

**Comment by Ditiae** [comment]
I think for comparing time input and actual benefit, #2 would be best. However, #3 would be pretty cool and hey, learning experience. Can always look into #3 later on too.

**Comment by EZtouch** [comment]
Here are my thoughts:

> (1) Simply change validation methods and wording of inputs

The user can see the expected input in the parentheses, and use the appropriate option.

> (2) Create some kind of file (perhaps configurations.txt) or something similar

The user will have to modify the file whenever a new bar needs to be created. The user will probably not be using the same configuration twice, since it will produce the same bar. If the user changes the syntax in the file by mistake and saves it, we'd have to recreate the file, and implement sophisticated validation for that. I believe this is not worthwhile, as it adds more complexity for the user and developer.

> (3) Create a UI to collect input, this allows for self-validation and of course is much more tidy for the user.

I am not very knowledgeable in how Python works, however, in other languages you could use a modular approach for this. By separating the interface from the logic, we can have three modules. The CLI, which interacts with the API, and the UI module, which also interacts with the API. This way, we can have the best of both worlds. 🙂👍

**Comment by NightShadeI** [comment]
Personally I am happy with option 3, however I think we should implement option 2 as a temporary solution, as I already have some of the required code from previous projects to do this and I believe it will work better than 1. Overtime, since python isn't nice when working with UI, I believe we can implement option 3 - since I will need to learn this from scratch (and I assume most people will here too).

EDIT: Should be able to get this done by tomorrow (option 2)

**Comment by Ditiae** [comment]
Option 2 I think would be great, if people decide to do small changes then they won't need to re-type every single option to test out one little extra thing.

**Comment by imnotaprogrammer** [comment]
On top of this, we should assert that the user enters magic abilities for magic attack style, ranged abilities for ranged, melee for melee. Universal abilities (like sacrifice, tuska's wrath) should be applicable to all 3 attack styles.

I have briefly explained this further and provided a suggestion on how the code should be structured moving forward in #14.

@EZtouch 

> I am not very knowledgeable in how Python works, however, in other languages you could use a modular approach for this. By separating the interface from the logic, we can have three modules. The CLI, which interacts with the API, and the UI module, which also interacts with the API. This way, we can have the best of both worlds.

Yes, I think using the MVC paradigm is our best bet here. MVC works with anything, it's just a fundamental stylistic choice for how to develop something, whether that's in Python, Java, C, or whatever, it'll all work :)
